### DISCUSSION TITLE
How to download a file attachment with golang sdk

### DISCUSSION THREAD

**Discussion by arnaudlemaignen** [microsoftgraph/msgraph-sdk-go#443] [General]
Hi 
I dont find anywhere examples on how downloading a mail attachment using msgraph-sdk-go v0.56.0

I have access to the attachment as followed
attachment,err := client.UsersById(emailUser).MessagesById(msgId).AttachmentsById(attId).Get(context.Background(), nil)

but then I dont see any methods to extract the content of the attachment.

how can this be achieved?
many thanks in advance

**Comment by baywet** [comment]
This is a limitation of how the API is designed today.
Effectively attachments can be fileAttachment, itemAttachment (contact card, meeting ics...) or referenceAttachment (another entity on Microsoft Graph) and only the former can be downloaded.
The service should allow you to specific an OData type cast segment in the URI to get access to download capabilities like so.

```
https://graph.microsoft.com/v1.0/users/userId/messages/messageId/attachments/microsoft.graph.fileAttachment/attachmentId/$value
```

This would roughly translate to this on the Go SDK

```go
attachmentValue, err := client.UsersById(emailUser).MessagesById(msgId).Attachments().GraphFileAttachmentById(attId).Value().Get(context.Background(), nil)
```
But because this isn't modeled properly, it's not working as expected.
The best workaround for the time being is to build the request URI yourself and use another request builder that performs a similar function, like

```go
rawUrl := "https://graph.microsoft.com/v1.0/users/userId/messages/messageId/attachments/attachmentId/$value"
builder := users.NewItemPhotoValueContentRequestBuilder(rawUrl, requestAdapter)
attachmentValue, err := builder.Get(context.Background(), nil)
```

https://github.com/microsoftgraph/msgraph-sdk-go/blob/c2137cc334f18950034177de1a7c8f414f79dd25/users/item_photo_value_content_request_builder.go#L46

**Reply by arnaudlemaignen** [reply]
Many thanks for your detailed answer, your WA is working fine !

Do you know when the proper model will be implemented (at least for fileAttachment) ?

**Reply by baywet** [reply]
No ETA to share on this, the best course of action would be to [create a feature request](https://feedbackportal.microsoft.com/feedback/forum/ebe2edae-97d1-ec11-a7b5-0022481f3c80) and get enough adhesion to it so the service team prioritizes this change accordingly.
